src/core/algorithms/dd/dd_verifier/dd_verifier.cpp

```diff
@@ -0,0 +1,211 @@
+        auto const &pair = hl.GetPairRows();
+        LOG(DEBUG) << "DD not Holds in " << col_schema->GetColumn(hl.GetAttributeIndex())->GetName()
+                   << " in " << pair.first << " and " << pair.second << " rows with values "
+                   << col_data.GetDataAsString(pair.first) << ", "
+                   << col_data.GetDataAsString(pair.second) << '\n';
+    }
+}
+
+std::size_t DDVerifier::GetNumErrorRhs() const {
+    return num_error_rhs_;
+}
+
+bool DDVerifier::IsColumnMetrizable(model::ColumnIndex const column_index) const {
+    model::TypedColumnData const &column = typed_relation_->GetColumnData(column_index);
+    model::TypeId const type_id = column.GetTypeId();
+
+    if (type_id == +model::TypeId::kUndefined) {
+        throw std::invalid_argument("Column with index \"" + std::to_string(column_index) +
+                                    "\" type undefined.");
+    }
+    if (type_id == +model::TypeId::kMixed) {
+        throw std::invalid_argument("Column with index \"" + std::to_string(column_index) +
+                                    "\" contains values of different types.");
+    }
+    return column.GetType().IsMetrizable();
+}
+
+double DDVerifier::CalculateDistance(model::ColumnIndex const column_index,
+                                     std::pair<std::size_t, std::size_t> const &tuple_pair) const {
+    double dif = 0;
+    model::TypedColumnData const &column = typed_relation_->GetColumnData(column_index);
+    std::byte const *first_value = column.GetValue(tuple_pair.first);
+    std::byte const *second_value = column.GetValue(tuple_pair.second);
+    auto const &type = static_cast<model::IMetrizableType const &>(column.GetType());
+    dif = type.Dist(first_value, second_value);
+    return dif;
+}
+
+std::vector<std::pair<int, int>> DDVerifier::GetRowsWhereLhsHolds() const {
+    std::vector<std::pair<int, int>> result;
+    std::vector<model::ColumnIndex> columns;
+    for (auto const &constraint : dd_.left) {
+        model::ColumnIndex column_index =
+                typed_relation_->GetSchema()->GetColumn(constraint.column_name)->GetIndex();
+        columns.push_back(column_index);
+    }
+    auto curr_constraint = dd_.left.cbegin();
+    for (std::size_t i = 0; i < num_rows_; i++) {
+        for (std::size_t j = i + 1; j < num_rows_; j++) {
+            if (auto const dif = CalculateDistance(columns[0], {i, j});
+                curr_constraint->constraint.Contains(dif)) {
+                result.emplace_back(i, j);
+            }
+        }
+    }
+    ++curr_constraint;
+    for (std::size_t i = 1; i < columns.size(); i++) {
+        std::vector<std::pair<int, int>> new_result;
+        for (std::size_t j = 0; j < result.size(); j++) {
+            if (double const dif = CalculateDistance(columns[i], result[j]);
+                curr_constraint->constraint.Contains(dif)) {
+                new_result.emplace_back(result[j]);
+            }
+        }
```

Comment:
We don't need value `j`, so let's just use range-based for:
```suggestion
        for (const auto& pair : result) {
            if (double const dif = CalculateDistance(columns[i], pair);
                curr_constraint->constraint.Contains(dif)) {
                new_result.emplace_back(pair);
            }
        }
```